File: 05-dynamic-programming/02_coin_change/coin_change.py
"""
Problem: Coin Change
LeetCode: #322 (Medium)

Time Complexity: O(Amount * N)
Space Complexity: O(Amount)
"""

from typing import List

def coin_change(coins: List[int], amount: int) -> int:
    dp = [float('inf')] * (amount + 1)
    dp[0] = 0

    for i in range(1, amount + 1):
        for coin in coins:
            if coin <= i:
                dp[i] = min(dp[i], dp[i - coin] + 1)

    return dp[amount] if dp[amount] != float('inf') else -1

if __name__ == "__main__":
    print("[Python] Coin Change Test")
    tests = [
        ([1, 2, 5], 11, 3),
        ([2], 3, -1),
        ([1], 0, 0),
    ]
    for coins, amount, expected in tests:
        res = coin_change(coins, amount)
        print(f"coins: {coins}, amount: {amount} -> Min Coins: {res} | Expected: {expected}")
        assert res == expected
    print("Coin change tests passed!")

File: 05-dynamic-programming/02_coin_change/coin_change.cpp
/**
 * Problem: Coin Change
 * LeetCode: #322 (Medium)
 * 
 * Time Complexity: O(Amount * N)
 * Space Complexity: O(Amount)
 */

#include <iostream>
#include <vector>
#include <algorithm>

int coinChange(const std::vector<int>& coins, int amount) {
    std::vector<int> dp(amount + 1, amount + 1);
    dp[0] = 0;

    for (int i = 1; i <= amount; ++i) {
        for (int coin : coins) {
            if (coin <= i) {
                dp[i] = std::min(dp[i], dp[i - coin] + 1);
            }
        }
    }
    return dp[amount] > amount ? -1 : dp[amount];
}

int main() {
    std::cout << "[C++] Coin Change Test\n";
    std::vector<int> coins = {1, 2, 5};
    int amount = 11;
    std::cout << "Min coins for " << amount << ": " << coinChange(coins, amount) << " (Expected: 3)\n";
    return 0;
}
